nav2_mppi_controller/src/noise_generator.cpp

```diff
@@ -69,26 +69,86 @@ void NoiseGenerator::setNoisedControls(
 {
   std::unique_lock<std::mutex> guard(noise_lock_);
 
+  computeAdaptiveStds(state);
+
   state.cvx = noises_vx_.rowwise() + control_sequence.vx.transpose();
   state.cvy = noises_vy_.rowwise() + control_sequence.vy.transpose();
   state.cwz = noises_wz_.rowwise() + control_sequence.wz.transpose();
 }
 
-void NoiseGenerator::reset(mppi::models::OptimizerSettings & settings, bool is_holonomic)
+void NoiseGenerator::computeAdaptiveStds(const models::State & state)
+{
+  auto & s = settings_;
+
+  // Should we apply decay function? or Any constraint is invalid?
+  if (s.advanced_constraints.wz_std_decay_strength <= 0.0f || !validateWzStdDecayConstraints()) {
+    wz_std_adaptive = s.sampling_std.wz;  // skip calculation
+  } else {
+    float current_speed;
+    if (is_holonomic_) {
+      const auto vx = static_cast<float>(state.speed.linear.x);
+      const auto vy = static_cast<float>(state.speed.linear.y);
+      current_speed = hypotf(vx, vy);
+    } else {
+      current_speed = std::fabs(static_cast<float>(state.speed.linear.x));
+    }
+
+    static const float e = std::exp(1.0f);
+    const float decayed_wz_std =
+      (s.sampling_std.wz - s.advanced_constraints.wz_std_decay_to) *
+      powf(e, -1.0f * s.advanced_constraints.wz_std_decay_strength * current_speed) +
+      s.advanced_constraints.wz_std_decay_to;
+
+    wz_std_adaptive = decayed_wz_std;
+  }
+}
+
+bool NoiseGenerator::validateWzStdDecayConstraints() const
 {
-  settings_ = settings;
-  is_holonomic_ = is_holonomic;
+  const models::AdvancedConstraints & c = settings_.advanced_constraints;
+  // Assume valid if angular decay is disabled
+  if (c.wz_std_decay_strength <= 0.0f) {
+    return true;  // valid
+  }
 
+  if (c.wz_std_decay_to < 0.0f || c.wz_std_decay_to > settings_.sampling_std.wz) {
+    return false;
+  }
+  return true;
+}
+
+float NoiseGenerator::getWzStdAdaptive() const
+{
+  return wz_std_adaptive;
+}
+
+void NoiseGenerator::reset(const mppi::models::OptimizerSettings & settings, bool is_holonomic)
+{
   // Recompute the noises on reset, initialization, and fallback
   {
     std::unique_lock<std::mutex> guard(noise_lock_);
+    // Copy settings and holonomic info after lock is acquired,
+    // otherwise we may encounter concurrent access errors
+    settings_ = settings;
+    is_holonomic_ = is_holonomic;
+    // reset initial adaptive value to parameterized value
+    wz_std_adaptive = settings_.sampling_std.wz;
     noises_vx_.setZero(settings_.batch_size, settings_.time_steps);
     noises_vy_.setZero(settings_.batch_size, settings_.time_steps);
     noises_wz_.setZero(settings_.batch_size, settings_.time_steps);
+
+    // Validate decay function, print warning message if decay_to is out of bounds
+    if (!validateWzStdDecayConstraints()) {
```

Comment:
If its invalid, throw an exception instead.

Actually if you do that then you can remove the Logger / node altogether.